Consolidate approach for conditional tests that check for privileged processes

jeffhandley opened the pull request:
Spawned from a [PR comment](https://github.com/dotnet/runtime/pull/77847/files#r1014171307) by @stephentoub in [Use Environment.IsPrivilegedProcess #77847](https://github.com/dotnet/runtime/pull/77847)

jeffhandley committed:
Consolidate approach for conditional tests that check for privileged processes

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -73,6 +73,8 @@ public static partial class PlatformDetection
         public static bool IsNotWindows => !IsWindows;
 
+        public static bool IsPrivilegedProcess => IsUnixAndSuperUser || IsWindowsAndElevated;
+
```
Do we still need the split across Unix/Windows, or can we view just have the implementation for this in the shared file and delete the OS specialization?

jeffhandley replied:
Refactored per that suggestion; thanks.

jeffhandley committed:
Simpler privileged process helpers

jeffhandley committed:
Revert SLN file changes

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Unix.cs:
```diff
@@ -55,9 +55,8 @@ public static partial class PlatformDetection
-
-        public static bool IsUnixAndSuperUser => !IsWindows && IsSuperUser;
+        public static bool IsSuperUser => IsBrowser || IsWindows ? false : IsPrivilegedProcess;
+        public static bool IsUnixAndSuperUser => !IsWindows && IsPrivilegedProcess;
```
Do we still need these? (Same question for the helpers in the Windows.cs file).

My expectation was that all tests could just use e.g.
```C#
[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsPrivilegedProcess))]
```
and if they also needed to only run on a specific platform (or be excluded from a specific platform), also use the standard:
```C#
[PlatformSpecific(...)]
```
mechanism.

Also, if the `IsBrowser` guard is necessary, that brings into question "why", and whether that means there's something wrong with IsPrivilegedProcess in the browser.

jeffhandley replied:
Thanks for clarifying what you were expecting--a deeper cut to remove the platform-specific helpers entirely, not just to refactor them. I couldn't see any evidence of why `IsBrowser` was necessary, but I anticipate some checks will fail and the reasoning will be revealed. If that does occur, I can then add `[SkipOnPlatform]` attributes where appropriate.

jeffhandley committed:
Remove platform-specific helpers for privileged process conditions

jeffhandley committed:
Fix System.IO tests for unix-specific privileged process scenarios

jeffhandley committed:
Browser does not support privileged processes

jeffhandley commented:
@stephentoub This should be ready for re-review. I've addressed the failing checks and found that on browser there's no concept of a privileged process, so that's now accounted for in `AdminHelpers.IsProcessElevated()`.

jeffhandley committed:
Fix System.IO tests for unix-specific privileged process scenarios

stephentoub commented:
> I've addressed the failing checks and found that on browser there's no concept of a privileged process

Shouldn't Environment.IsPrivilegedProcess just return false on browser then?

jeffhandley committed:
Enable the Environment.IsPrivilegedProcess test for browser

jeffhandley commented:
> Shouldn't Environment.IsPrivilegedProcess just return false on browser then?

It does. [`Environment.IsPrivilegedProcess`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Environment.cs#L26-L37) calls the platform-specific [`IsPrivilegedProcessCore`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs#L28), which just returns `false`.

I also walked the blame back for the affected tests and found that when they were being enabled for WASM, that's when the `IsBrowser` conditions were added to short-circuit to `false` despite satisfying `!IsWindows`.

I did just find a unit test that was being skipped on browser that can now be enabled since `AdminHelpers.IsProcessElevated()` will now return `false` on browser instead of throwing.

stephentoub commented:
> It does. [Environment.IsPrivilegedProcess](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Environment.cs#L26-L37) calls the platform-specific [IsPrivilegedProcessCore](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs#L28), which just returns false.

Ah, I forgot that AdminHelpers.IsProcessElevated is reimplementing the equivalent of IsPrivilegedProcess.

Thanks.

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Windows.cs:
```diff
@@ -248,31 +248,8 @@ public static bool IsInAppContainer
         }
 
-        public static bool CanRunImpersonatedTests => PlatformDetection.IsNotWindowsNanoServer && PlatformDetection.IsWindowsAndElevated;
+        public static bool CanRunImpersonatedTests => PlatformDetection.IsNotWindowsNanoServer && PlatformDetection.IsWindows && PlatformDetection.IsPrivilegedProcess;
```
Will PlatformDetection.IsWindows ever be false here?

jeffhandley replied:
Yes. Despite the `.Windows.cs`, file name, the file is unconditionally compiled so Windows-specific helpers like this are defined similarly throughout this file.

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -73,6 +73,22 @@ public static partial class PlatformDetection
         public static bool Is64BitProcess => IntPtr.Size == 8;
         public static bool IsNotWindows => !IsWindows;
 
+        private static int s_isPrivilegedProcess = -1;
```
```suggestion
        private static volatile int s_isPrivilegedProcess = -1;
```

jeffhandley replied:
Today I learned. Thanks!

stephentoub commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -73,6 +73,22 @@ public static partial class PlatformDetection
+        {
+            get
+            {
+                if (s_isPrivilegedProcess != -1)
+                    return s_isPrivilegedProcess == 1;
+
+                s_isPrivilegedProcess = AdminHelpers.IsProcessElevated() ? 1 : 0;
+
+                return s_isPrivilegedProcess == 1;
```
```suggestion
                int p = s_isPrivilegedProcess;
                if (p == -1)
                {
                    s_isPrivilegedProcess = p = AdminHelpers.IsProcessElevated() ? 1 : 0;
                }
                
                return p == 1;
```

stephentoub commented on src/libraries/System.Formats.Tar/tests/TarReader/TarReader.TarEntry.ExtractToFile.Tests.Unix.cs:
```diff
@@ -9,8 +9,10 @@ namespace System.Formats.Tar.Tests
     {
-        [PlatformSpecific(TestPlatforms.AnyUnix & ~TestPlatforms.tvOS)] // https://github.com/dotnet/runtime/issues/68360
-        [ConditionalFact(nameof(IsUnixButNotSuperUser), nameof(IsNotLinuxBionic))]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
```
In the above .Unix.cs file there's no PlatformSpecific on the test... I assumed that was because that .Unix.cs file is only compiled into unix builds, but then here this is also a .Unix.cs and this one does have PlatformSpecific...?

stephentoub commented on src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.Windows.cs:
```diff
@@ -132,8 +133,8 @@ public NamedPipeTest_CurrentUserOnly_Windows(TestAccountImpersonator testAccount
 
-        [OuterLoop]
-        [ConditionalFact(nameof(IsAdminOnSupportedWindowsVersions))]
+        [OuterLoop("Requires admin privileges")]
```
Is this the reason it's OuterLoop?  That's acovered by the ConditionalFact condition, and there are other tests that require admin privileges that aren't OuterLoop.

stephentoub commented on src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs:
```diff
@@ -793,13 +793,13 @@ public async Task Ping_TimedOut_TAP_Success()
         [InlineData(AddressFamily.InterNetwork)]
         [InlineData(AddressFamily.InterNetworkV6)]
-        [OuterLoop] // Depends on sudo
+        [OuterLoop("Requires sudo access")]
```
Same here; seems a strange reason to make a test outerloop, since due to the condition above we'll run this test if the process is privileged.

stephentoub reviewed (approved):
Thanks!

jeffhandley committed:
PR feedback

Omit [PlatformSpecific] in modified tests that are in platform-specific test files.

Remove [Outerloop] attributes from privileged process tests.

Better multithreaded handling in PlatformDetection.IsPrivilegedProcess

jeffhandley commented:
MacCatalyst failure is #78778
